### PR TITLE
[5.2][CodeCompletion] Use GenericSignature methods to get 'associatedtype' requirements

### PR DESCRIPTION
**Pull request by rintaro**
Cherry-pick of #29086 into `swift-5.2-branch` originally reviewed by @slavapestov 

rdar://problem/57245073

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
The added test case depends on https://github.com/apple/swift/pull/29104. Holding off until it's merged

**Review by DougGregor — APPROVED** [review]

**Pull request merged**
